PEP 701: Syntactic formalization of f-strings (#2907)
python/peps: pep-0701.txt

        Batuhan Taskaya <batuhan@python.org>,
        Lysandros Nikolaou <lisandrosnik@gmail.com>
Discussions-To: discuss.python.org
Type: Standards Track
Status: Draft
Content-Type: text/x-rst
Python-Version: 3.12
Created: 15-Nov-2022

Abstract
========

This document proposes to lift some of the restrictions originally formulated in
:pep:`498` and to provide a formalized grammar for f-strings that can be
integrated into the parser directly. The proposed syntactic formalization of
f-strings will have some small side-effects over how f-strings are parsed and
interpreted, allowing for a considerable number of advantages for end users and
library developers, while also improving dramatically the maintainance cost of
the code dedicated to parse f-strings.

Motivation
==========

When f-strings were originally introduced in :pep:`498` the specification was
provided without providing a formal grammar for f-strings. Additionally, the
specification contains several restrictions that are imposed so the parsing of
f-strings could be implemented into CPython without modifying the existing
lexer. These limitations have been recognized previously and previous attempts

> have been made to lift them in :pep:`536` but non of this work was ever implemented
> (see [2]_). Some of these limitations (collected originally by :pep:`536`) are:

Reviewer: ```suggestion
have been made to lift them in :pep:`536`, but `none of this work was ever implemented`_.
Some of these limitations (collected originally by :pep:`536`) are:
```

Use actual reST link, fix typo and add missing comma

#. It is impossible to use the quote character delimiting the f-string
   within the expression portion::

    >>> f'Magic wand: { bag['wand'] }'
                                 ^
    SyntaxError: invalid syntax

#. A previously considered way around it would lead to escape sequences
   in executed code and is prohibited in f-strings::

    >>> f'Magic wand { bag[\'wand\'] } string'
    SyntaxError: f-string expression portion cannot include a backslash

#. Comments are forbidden even in multi-line f-strings::

    >>> f'''A complex trick: {
    ... bag['bag']  # recursive bags!
    ... }'''
    SyntaxError: f-string expression part cannot include '#'

#. Arbitrary nesting of expressions without expansion of escape sequences is
   available in every single other language employing a string interpolation

>    method that uses expressions instead of just variable names. [6]_

Reviewer: ```suggestion
   method that uses expressions instead of just variable names, `per Wikipedia`_.
```

Make this an actual direct link, rather than an indirect footnote

CPython relies on tokenising f-strings as `STRING` tokens and a post processing of
these tokens. This has the following problems:

#. It adds a considerable maintainance cost to the CPython parser. This is because
   the parsing code needs to be written by hand, which has historically lead to a
   considerable number of inconsistencies and bugs. Writing and maintaining parsing
   code by hand in C has always been considered error prone and dangerous as it needs
   to deal with a lot of manual memory management over the original lexer buffers.

#. The f-string parsing code is not able to use the new improved error message mechanisms
   that the new PEG parser, originally introduced in :pep:`617`, has allowed. The
   improvements that these error messages brought has been greatly celebrated but
   unfortunately f-strings cannot benefit from them because they are parsed in a
   separate piece of the parsing machinery. This is especially unfortunate, since
   there are several syntactical features of f-strings that can be confusing due
   to the different implicit tokenization that happens inside the expression
   part (for instance ``f"{y:=3}"`` is not an assignment expression).

> #. Other Python implementations have no way to know if they have implemented

Reviewer: Resolving this for now so it doesn't block merge.

   f-strings correctly because contrary to other language features, they are not

>    part of the official Grammar [1]_. This is important because several prominent

Reviewer: ```suggestion
   part of the :ref:`official Python grammar <f-strings>`.
   This is important because several prominent
```

Use a direct Intersphinx reference rather than a manual hardcoded URL

>    alternative implementations such as PyPy are using CPython's PEG parser
>    (see [3]_) and/or are basing their Grammars on the official PEG Grammar. The

Reviewer: ```suggestion
   alternative implementations are using CPython's PEG parser, `such as PyPy`_,
   and/or are basing their grammars on the official PEG Grammar. The
```

Link this directly, and use correct capitalization for a common noun

   fact that f-strings use a separate parser prevents these alternative implementations
   to leverage the official grammar and to benefit to improvements in error messages derived
   from the grammar.

> A version of this proposal was originally discussed in Python dev [4]_ and

Reviewer: ```suggestion
A version of this proposal was originally `discussed on Python-Dev`_  and
```

Fix grammar and use actual reST link

> presented at the Python language summit 2022 [5]_ where it was enthusiastically

Reviewer: ```suggestion
`presented at the Python Language Summit 2022`_ where it was enthusiastically
```

Make this an actual link

no fundamental need for users to learn anything new. However, as the formalized grammar
allows some new possibilities, it is important that the formal grammar is added to the
documentation and explained in detail, explicitly mentioning what constructs are possible
since this PEP is aiming to avoid confusion.

It is also beneficial to provide users with a simple framework for understanding what can
be placed inside an f-string expression. In this case the authors think that this work will
make even more simple to explain this aspect of the language since it can be summarized as:

    "You can place any valid Python expression inside an f-string expression".

With the changes in this PEP, there is no need to clarify that string quotes are
limited to be different as the quotes of the enclosing string because this is
now allowed: as an arbitrary Python string can contain any possible choice of
quotes, so can any f-string expression. Additionally there is no need to clarify
that certain things are not allowed in the expression part because of
implementation restructions such as comments, new line characters or
backslashes. 

> The only "surprising" difference is that as the f-string allows to specificy a

Reviewer: ```suggestion
The only "surprising" difference is that as f-strings allow specifying a
```

Fix typo and grammar

> format, expressions that allow a ``:`` character at top level still need to be

Reviewer: ```suggestion
format, expressions that allow a ``:`` character at the top level still need to be
```

Fix grammar

> enclosed in parenthesis. This is not new to this work, but is important to

Reviewer: ```suggestion
enclosed in parenthesis. This is not new to this work, but it is important to
```

Fix grammar

emphasize that this restriction is still in place. This allows for an easier
modification of the summary:

    You can place any valid Python expression inside

>     an f-string expression, and everything after a ``:`` character at top level will

Reviewer: ```suggestion
    an f-string expression, and everything after a ``:`` character at the top level will
```

Grammar

>     be identified as a format specification

Reviewer: ```suggestion
    be identified as a format specification.
```

Add period, to be consistent with previous directly parallel quoted sentence

Reference Implementation
========================

A reference implementation can be found in the implementation_ fork.

Rejected Ideas
==============

#. We have decided not to lift the restriction that some expression portions

>    need to wrap ``':'`` and ``'!'`` in braces at top level, e.g.::

Reviewer: ```suggestion
   need to wrap ``':'`` and ``'!'`` in braces at the top level, e.g.::
```

Fix grammar

    >>> f'Useless use of lambdas: { lambda x: x*2 }'
    SyntaxError: unexpected EOF while parsing
   
   The reason is that this would this will introduce a considerable amount of
   complexity for no real benefit. This is due to the fact that the ``:`` character
   normally separates the f-string format specification. This format specification
   is currently tokenized as a string. As the tokenizer MUST tokenize what's on the

>    right of the ``:`` as either a string or a stream of tokens this won't allow the

Reviewer: ```suggestion
   right of the ``:`` as either a string or a stream of tokens, this won't allow the
```

Particularly critical missing comma in a run-on sentence

   parser to differentiate between the different semantics as that would require the
   tokenizer to backtrack and produce a different set of tokens (this is, first try

>    as a stream of tokens and if it fails try as a string for a format specifier).

Reviewer: ```suggestion
   as a stream of tokens, and if it fails, try as a string for a format specifier).
```

Missing commas

   As there is no fundamental advantage in being able to allow lambdas and similar

>    expressions at top level, we have decided to keep the restriction that these must

Reviewer: ```suggestion
   expressions at the top level, we have decided to keep the restriction that these must
```

Fix grammar

   be parenthesized if needed::

    >>> f'Useless use of lambdas: { (lambda x: x*2) }'
  

Open Issues
===========

> [Any points that are still being decided/discussed.]

Reviewer: If there are none yet, replace this placeholder with:

```suggestion
None yet
```

Footnotes
=========

> .. [1] "Grammar for f-strings":
>     https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals
> 
> .. [2] "Deferred notice on PEP 536"
>     https://mail.python.org/archives/list/python-dev@python.org/thread/N43O4KNLZW4U7YZC4NVPCETZIVRDUVU2/#NM2A37THVIXXEYR4J5ZPTNLXGGUNFRLZ
> 
> .. [3] "Pypy uses now CPython's PEG parser"
>     https://foss.heptapod.net/pypy/pypy/-/commit/fe120f89bf07e64a41de62b224e4a3d80e0fe0d4/pipelines?ref=branch%2Fpy3.9
> 
> .. [4] "Python-dev discussion about f-strings in the grammar"
>     https://mail.python.org/archives/list/python-dev@python.org/thread/54N3MOYVBDSJQZTU6MTCPLUPIFSDN5IS/#SAYU6SMP4KT7G7AQ6WVQYUDOSZPKHJMS
> 
> .. [5] "Language summit 2022"
>     https://pyfound.blogspot.com/2022/05/the-2022-python-language-summit-f.html
> 
> .. [6] Wikipedia article on string interpolation
>     https://en.wikipedia.org/wiki/String_interpolation

Reviewer: Per #2130 and [PEP 12 guidance](https://peps.python.org/pep-0012/#hyperlinks), actual reST links should be used instead of footnotes for URLs, as it forces the reader through an extra layer of indirection for little benefit.

Also [per PEP 12](https://peps.python.org/pep-0012/#footnotes), if you do use footnotes, named auto-numbered footnotes should always be used instead of hardcoded numbers, as it is easy for the numbers to get out of sync with the footnotes they reference, is tedious to manually renumber them and it ensures the numbers appear in the same order they are first used in the file (currently, they are already well out of order).

Finally, I replaced one hardcoded URL with a direct Intersphinx reference, and the Wikipedia one to link directly to the section being referenced.

```suggestion
.. _official Python grammar: https://docs.python.org/3/reference/lexical_analysis.html#formatted-string-literals

.. _none of this work was ever implemented: https://mail.python.org/archives/list/python-dev@python.org/thread/N43O4KNLZW4U7YZC4NVPCETZIVRDUVU2/#NM2A37THVIXXEYR4J5ZPTNLXGGUNFRLZ

.. _such as PyPy: https://foss.heptapod.net/pypy/pypy/-/commit/fe120f89bf07e64a41de62b224e4a3d80e0fe0d4/pipelines?ref=branch%2Fpy3.9

.. _discussed on Python-Dev: https://mail.python.org/archives/list/python-dev@python.org/thread/54N3MOYVBDSJQZTU6MTCPLUPIFSDN5IS/#SAYU6SMP4KT7G7AQ6WVQYUDOSZPKHJMS

.. _presented at the Python Language Summit 2022: https://pyfound.blogspot.com/2022/05/the-2022-python-language-summit-f.html

.. _per Wikipedia: https://en.wikipedia.org/wiki/String_interpolation#Examples
```
